### ISSUE TITLE
Tools: Rule backfilling support in Thanos

### ISSUE THREAD

**Issue by yeya24** [thanos-io/thanos#5776]
### Is your proposal related to a problem?

`promtool` has supported the rule backfilling feature already, which does rule evaluation using standard Prometheus query API and generates TSDB blocks.
I have tried it before to do backfilling in my usecase. However, Thanos's query API supports some additional parameters like `max_source_resolution`. Promtool doesn't support this so I had to modify its code.

Besides, promtool doesn't support uploading data to the object storage. So to upload data we have to use it along with the thanos sidecar. It would be handy to support that feature.

### Describe the solution you'd like

Add a new Thanos subcommand to run recording rules backfilling.

### Describe alternatives you've considered

Support these Thanos specific params in Promtool directly.

**Comment by TomHellier** [comment]
I'm looking for something similar, but was able to use the existing tooling.

Using promtool, I was able to connect to `thanos-query`, and run recording rules against historical data. This output some TSDB files into a directory.

I then used [thanos tool bucket-upload](https://thanos.io/tip/components/tools.md/#bucket-upload-blocks) which was added in #6263 which let's you upload TSDBs created from promtool direct to an object store, providing some external labels.

It took a while to show up in the Announced LabelSets of the `thanos-storegateways`, but did eventually appear.

**Comment by yeya24** [comment]
Yeah it could work but in my case the data requires specifying `max_source_resolution` because the data months ago were already downsampled. There is no way to specify that in promtool.
